File: app/src/main/java/com/ryanrvldo/moviecatalogue/viewmodel/RecentSearchViewModel.java
package com.ryanrvldo.moviecatalogue.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.ryanrvldo.moviecatalogue.data.Repository;
import com.ryanrvldo.moviecatalogue.data.source.model.Search;

import java.util.List;

public class RecentSearchViewModel extends ViewModel {
    private Repository repository;
    private LiveData<List<Search>> searchQuery;

    public RecentSearchViewModel(Repository repository) {
        this.repository = repository;
    }

    public LiveData<List<Search>> getSearchQuery() {
        if (searchQuery == null) {
            searchQuery = new MutableLiveData<>();
            searchQuery = repository.getSearchHistory();
        }
        return searchQuery;
    }

    public void insertSearch(Search search) {
        repository.insertSearch(search);
        searchQuery = repository.getSearchHistory();
    }

    public void deleteSearch(Search search) {
        repository.deleteSearch(search);
        searchQuery = repository.getSearchHistory();
    }

    public void deleteAllSearch() {
        repository.deleteAllSearchHistory();
        searchQuery = repository.getSearchHistory();
    }

    public String selectSearch(String query) {
        return repository.getSearch(query);
    }
}

File: app/src/main/java/com/ryanrvldo/moviecatalogue/ui/viewmodel/RecentSearchViewModel.kt
package com.ryanrvldo.moviecatalogue.ui.viewmodel

import androidx.hilt.lifecycle.ViewModelInject
import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ryanrvldo.moviecatalogue.data.Repository
import com.ryanrvldo.moviecatalogue.data.model.Search
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class RecentSearchViewModel @ViewModelInject constructor(
    private val repository: Repository
) : ViewModel() {

    val searchHistories: LiveData<List<Search>> by lazy {
        repository.getSearchHistories()
    }

    fun addSearchQuery(search: Search) = viewModelScope.launch(Dispatchers.IO) {
        repository.addSearchQuery(search)
    }

    fun removeSearchQuery(search: Search) = viewModelScope.launch(Dispatchers.IO) {
        repository.removeSearchQuery(search)
    }

    fun removeSearchHistories() = viewModelScope.launch(Dispatchers.IO) {
        repository.removeSearchHistories()
    }

    suspend fun selectSearch(query: String): String? = repository.selectSearchQuery(query)
}